src/flp/gadgets.rs

```diff
@@ -438,25 +464,43 @@ where
     fn call_poly(&mut self, outp: &mut [F], inp: &[Vec<F>]) -> Result<(), FlpError> {
         gadget_call_poly_check(self, outp, inp)?;
 
+        // Create a copy of the inner gadget and two working buffers on each thread. Evaluate the
+        // gadget on each input polynomial, using the first temporary buffer as an output buffer.
+        // Then accumulate that result into the second temporary buffer, which acts as a running
+        // sum. Then, discard everything but the partial sums, add them, and finally copy the sum
+        // to the output parameter. This is equivalent to the single threaded calculation in
+        // ParallelSum, since we only rearrange additions, and field addition is associative.
         let res = inp
             .par_chunks(self.serial_sum.inner.arity())
-            .map(|chunk| {
-                let mut inner = self.serial_sum.inner.clone();
-                let mut partial_outp = vec![F::zero(); outp.len()];
-                inner.call_poly(&mut partial_outp, chunk).unwrap();
-                partial_outp
-            })
+            .fold(
+                || ParallelSumFoldState::new(&self.serial_sum.inner, outp.len()),
+                |mut state, chunk| {
+                    state
+                        .inner
+                        .call_poly(&mut state.partial_output, chunk)
+                        .unwrap();
+                    for (sum_elem, output_elem) in state
+                        .partial_sum
+                        .iter_mut()
+                        .zip(state.partial_output.iter())
+                    {
+                        *sum_elem += *output_elem;
+                    }
+                    state
+                },
+            )
+            .map(|state| state.partial_sum)
             .reduce(
                 || vec![F::zero(); outp.len()],
                 |mut x, y| {
-                    for i in 0..x.len() {
-                        x[i] += y[i];
+                    for (xi, yi) in x.iter_mut().zip(y.iter()) {
+                        *xi += *yi;
```

Comment:
There's so many places where we use `foo.iter_mut().zip(&bar)` where `foo` and `bar` are `Vec<F: FieldElement>` that I'm wondering if we should put an implementation of `std::ops::Add` for `Vec<F: FieldElement>` in `field.rs` and use it everywhere. That isn't needed in this PR, however.

Reply:
Sounds good to me, I'll open an issue.